agent/api/statechange.go

```diff
@@ -377,3 +459,129 @@ func (ts TaskStateChange) GetEventType() statechange.EventType {
+		ContainerName: aws.String(change.ContainerName),
+	}
+	if change.RuntimeID != "" {
+		statechange.RuntimeId = aws.String(change.RuntimeID)
+	}
+	if change.Reason != "" {
+		statechange.Reason = aws.String(change.Reason)
+	}
+	if change.ImageDigest != "" {
+		statechange.ImageDigest = aws.String(change.ImageDigest)
+	}
+
+	stat := change.Status.String()
+	if stat != apicontainerstatus.ContainerStopped.String() && stat != apicontainerstatus.ContainerRunning.String() {
+		logger.Warn("Not submitting unsupported upstream container state", logger.Fields{
+			field.Status:        stat,
+			field.ContainerName: change.ContainerName,
+			field.TaskARN:       change.TaskArn,
+		})
+		return nil, nil
+	}
+	if stat == "DEAD" {
+		stat = apicontainerstatus.ContainerStopped.String()
+	}
+	statechange.Status = aws.String(stat)
+
+	if change.ExitCode != nil {
+		exitCode := int64(aws.IntValue(change.ExitCode))
+		statechange.ExitCode = aws.Int64(exitCode)
+	}
+
+	networkBindings := getNetworkBindings(change)
+	// we enforce a limit on the no. of network bindings for containers with at-least 1 port range requested.
+	// this limit is enforced by ECS, and we fail early and don't call SubmitContainerStateChange.
+	if change.Container.HasPortRange() && len(networkBindings) > ecsMaxNetworkBindingsLength {
+		return nil, fmt.Errorf("no. of network bindings %v is more than the maximum supported no. %v, "+
+			"container: %s "+"task: %s", len(networkBindings), ecsMaxNetworkBindingsLength, change.ContainerName, change.TaskArn)
+	}
+	statechange.NetworkBindings = networkBindings
+
+	return statechange, nil
+}
+
+// ProtocolBindIP used to store protocol and bindIP information associated to a particular host port
+type ProtocolBindIP struct {
+	protocol string
+	bindIP   string
+}
+
+// getNetworkBindings returns the list of networkingBindings, sent to ECS as part of the container state change payload
+func getNetworkBindings(change ContainerStateChange) []*ecsmodel.NetworkBinding {
+	networkBindings := []*ecsmodel.NetworkBinding{}
+	// hostPortToProtocolBindIPMap is a map to store protocol and bindIP information associated to host ports
+	// that belong to a range. This is used in case when there are multiple protocol/bindIP combinations associated to a
+	// port binding. example: when both IPv4 and IPv6 bindIPs are populated by docker.
+	hostPortToProtocolBindIPMap := map[int64][]ProtocolBindIP{}
+
+	// ContainerPortSet consists of singular ports, and ports that belong to a range, but for which we were not able to
+	// find contiguous host ports and ask docker to pick instead.
+	containerPortSet := change.Container.GetContainerPortSet()
+	// each entry in the ContainerPortRangeMap implies that we found a contiguous host port range for the same
+	containerPortRangeMap := change.Container.GetContainerPortRangeMap()
+
+	for _, binding := range change.PortBindings {
+		hostPort := int64(binding.HostPort)
+		containerPort := int64(binding.ContainerPort)
```

Comment:
if one or more of the containerPorts has a zero default value, we'll end up hitting the `else` statement below multiple times with the same zero port.  Just want to be sure we're accounting for all the duplicate key scenarios that might come up here.

Reply:
if this is already accounted for, can you add a test below to assert it?

Reply:
We should not encounter a scenario where multiple port bindings' container port is zero. ECS Control Plane only considers values are between 1 and 65535 as valid for containerPortRange. We also have a check in place in our Agent to ensure that port zero is not included in the set of container ports that are exposed for a container here: https://github.com/aws/amazon-ecs-agent/blob/master/agent/api/task/task.go#L1822

Reply:
Discussed offline: I will add a unit test for the Agent code I linked in my previous reply to assert that port zero will explicitly not be included in the set of container ports that are exposed for a container.